src/abstract_domains.rs

```diff
@@ -244,6 +244,41 @@ pub enum ExpressionDomain {
 
     /// The corresponding concrete value is the runtime address of location identified by the path.
     Reference(Path),
+
+    /// The unknown value of a place in memory.
+    /// This is distinct from Top in that we known something: the place and the type.
+    /// This is a useful distinction because it allows us to simplify some expressions
+    /// like x == x. The type is needed to prevent this particular optimization if
+    /// the variable is a floating point number that could be NaN.
+    Variable {
+        path: Box<Path>,
+        var_type: ExpressionType,
+    },
+}
+
+/// The type of a place in memory, as understood by MIR.
+/// For now, we are only really interested to distinguish between
+/// floating point values and other values, because NaN != NaN.
+/// In the future the other distinctions may be helpful to SMT solvers.
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum ExpressionType {
+    Bool,
+    Char,
+    F32,
+    F64,
```

Comment:
Careful if F128 ends up getting in.

Reply:
The type checker will complain when that happens.